semantic/flatbuffers.go

```diff
@@ -556,3 +556,41 @@ func (p *FunctionParameter) FromBuf(fb *fbsemantic.FunctionParameter) error {
 	}
 	return nil
 }
+
+type fbTyper interface {
+	TypType() byte
+	Typ(obj *flatbuffers.Table) bool
+}
+
+// getMonoType produces an FBMonoType from the given FlatBuffers expression that has
+// a union "typ" field (which is all the different kinds of expressions).
+// To start with not all expressions may have types, so this will not return an error
+// if the type is missing---this shold change in the future.
+// TODO(cwolff): return an error if the type is missing
+func getMonoType(fbExpr fbTyper) (*FBMonoType, error) {
```

Comment:
So `getMonoType` takes a flatbuffer as input, checks to see if the current index is the start of a MonoType, and if so, returns it? Do I have this right?

Also the `Typ` method on the interface ensures we only call this on nodes that have a `Typ`?

Reply:
The code in `fbgen/cmd/semantic.go` looks at each struct type in the Go `semantic` package that implements `Node` and if it has a field of type `*FBMonoType` then it will generate code in `flatbuffers_gen.go` that calls this function. All such fields are accessed as `TypType()` and `Typ()` in the FlatBuffer schema. If this were not the case, then the Go code would not compile.

Essentially, instances of `fbTyper` are all the structs generated by `flatc` that are expressions.  This method just extracts the type from the flatbuffers expression (sadly the only common interface for it is `fbTabler`), and returns it bundled with its `fbsemantic.MonoType` tag.

Also I still need to copy the byte slice here to ensure we don't rely on the original underlying flatbuffer hanging around.